pkg/stanza/fileconsumer/internal/trie/trie_test.go

```diff
@@ -198,3 +198,359 @@ func runTest(t *testing.T, trie *Trie, tc *testCase) {
+												ops: []testOp{
+													del("ABCxyz", true, "just confirmed ABCxyz exists"),
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					"Found:ABCxyz,ABCDEFxyz": opTree{ // Next poll finds ABCxyz and ABCDEFxyz
+						ops: []testOp{
+							// Inserting longer string first
+							has("ABCDEFxyz", true, "recognize ABCDEF w/ xyz appended"),
+							put("ABCDEFxyz"), // TODO how do we know we need to call this?
+							has("ABCDEF", false, "ABCDEF should have been pushed down to ABCDEFxyz"),
+							has("ABC", true, "ABC should not have been affected"),
+
+							has("ABCxyz", true, "recognize ABC w/ xyz appended"),
+							put("ABCxyz"), // TODO how do we know we need to call this?
+							has("ABC", false, "ABC should have been pushed down to ABCxyz"),
+						},
+						continuations: map[string]opTree{
+							"Done:ABCxyz": opTree{ // Finish reading ABCxyz and remove from trie
+								ops: []testOp{
+									del("ABCxyz", true, "just confirmed ABCxyz exists"),
+									has("ABCDEFxyz", true, "ABCDEFxyz should not have been deleted"),
+								},
+								continuations: map[string]opTree{
+									"Done:ABCDEFxyz": opTree{ // Finish reading ABCDEFxyz and remove from trie
+										ops: []testOp{
+											del("ABCDEFxyz", true, "just confirmed ABCDEFxyz exists"),
+										},
+									},
+								},
+							},
+							"Done:ABCDEFxyz": opTree{ // Finish reading ABCDEFxyz and remove from trie
+								ops: []testOp{
+									del("ABCDEFxyz", true, "just confirmed ABCDEFxyz exists"),
+									has("ABCxyz", true, "ABCxyz should not have been deleted"),
+								},
+								continuations: map[string]opTree{
+									"Done:ABCxyz": opTree{ // Finish reading ABCxyz and remove from trie
+										ops: []testOp{
+											del("ABCxyz", true, "just confirmed ABCxyz exists"),
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}.run(t, []testOp{})(t)
+}
+
+// testOp is one HasKey, Put, or Delete call to the trie,
+// along with validation of expectations.
+type testOp func(t *testing.T, trie *Trie)
+
+func has(key string, expect bool, why string) testOp {
+	return func(t *testing.T, trie *Trie) {
+		assert.Equalf(t, trie.HasKey([]byte(key)), expect, why)
+	}
+}
+
+// put automatically asserts that the trie contains the key after adding.
+func put(key string) testOp {
+	return func(t *testing.T, trie *Trie) {
+		trie.Put([]byte(key))
+		assert.Truef(t, trie.HasKey([]byte(key)), "called Put(%s) but HasKey(%s) is still false", key, key)
+	}
+}
+
+// del automatically asserts that the trie no longer contains the key after deleting it.
+func del(key string, expect bool, why string) testOp {
+	return func(t *testing.T, trie *Trie) {
+		assert.Equalf(t, trie.Delete([]byte(key)), expect, why)
+		assert.Falsef(t, trie.HasKey([]byte(key)), "called Delete(%s) but HasKey(%s) is still true", key, key)
```

Comment:
Well, this won't necessarily be false.
Consider, `A-B-C-D(true)-E-F(true)`
Deleting `ABCDEF` will cause `A-B-C-D(true)`.
Now, if we do `HasKey(ABCDEF)`, it will return `true` coz `ABCD \w EF appended`

Reply:
Need now worry about losing the data, even if the file gets deleted, we have the reader stored.
Will close out readers after emitting which haven't been redectected in last 3 poll cycles.